working/macros/api/expansion_protocol.dart

```diff
@@ -0,0 +1,109 @@
+import 'package:package_config/package_config.dart';
+
+import 'builders.dart';
+import 'code.dart';
+import 'introspection.dart';
+
+/// The interface used by Dart language implementations, in order to load
+/// and execute macros, as well as produce library augmentations from those
+/// macro applications.
+///
+/// This class more clearly defines the role of a Dart language implementation
+/// during macro discovery and expansion, and unifies how augmentation libraries
+/// are produced, as well as how (and where) they are written to disk.
+abstract class MacroExecutor {
+  /// Creating an executor will require at least a package config, in order to
+  /// resolve macros by URI.
+  MacroExpander(PackageConfig packageConfig);
+
+  /// Invoked when an implementation discovers a new macro definition in a
+  /// library, and prepares this executor to run the macro.
+  ///
+  /// May be invoked more than once for the same macro, which will cause the
+  /// macro to be re-loaded. Previous [MacroClassIdentifier]s and
+  /// [MacroInstanceIdentifier]s given for this macro will be invalid after
+  /// that point and should be discarded.
+  ///
+  /// Throws an exception if the macro fails to load.
+  Future<MacroClassIdentifier> loadMacro(Uri library, String name);
```

Comment:
It would be any valid import uri that can be resolvable with the package config. This api is intentionally abstracting away the details - so for instance I would expect one implementation to load it via `IsolateMirror.loadUri`.

For testing we could have a specific implementation of this interface that uses an in memory file system. It would also allow you to unit test in a more black boxed way with this part of the system (you could have a mock `MacroExecutor` that never actually runs anything, just returns fake declarations).

Reply:
I created a very basic and partial implementation to hopefully help here https://github.com/dart-lang/language/pull/2022